### ISSUE TITLE
rustc macro: unexpected compilation error when `ty` designator is used within where clause

### ISSUE THREAD

**Issue by 0xBACD** [rust-lang/rust#97719]
This code doesn't seem to compile with the following error [[playground]](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=eca2db7fd82b9dd6c0dc00494cf72ff0):
```rust
macro_rules! foo {
    ($t:ty) => {
        fn bar<T>(_: T) where T: $t {}
    }
}

foo!(std::fmt::Debug);
```
```
error: expected one of `!`, `(`, `,`, `?`, `for`, `{`, `~`, lifetime, or path, found `std::fmt::Debug`
 --> src/main.rs:3:34
  |
3 |         fn bar<T>(_: T) where T: $t {}
  |                                  ^^ expected one of 9 possible tokens
...
7 | foo!(std::fmt::Debug);
  | --------------------- in this macro invocation
  |
  = note: this error originates in the macro `foo` (in Nightly builds, run with -Z macro-backtrace for more info)
```
There should be no error since `ty` designator stands for type, `std::fmt::Debug` is definitely a type and where clause accepts trait types.
This problem persists on any channel I've tested it - stable, beta and nightly.

**Comment by compiler-errors** [comment]
In this position, `std::fmt::Debug` is not meant to be a type. It's meant to be a trait, which is parsed differently than a type. 

Have you tried using the `path` macro fragment?

**Comment by 0xBACD** [comment]
Yeah, it seems I just misread the reference. Actually it says that `ty` only matches to trait objects and impl trait so compiler behaviour here seems legit. Thank you, @compiler-errors!

**Comment by petrochenkov** [comment]
FWIW, this will work once https://github.com/rust-lang/rust/pull/96724 and subsequent PRs land (maybe not very soon).
`std::fmt::Debug` will be treated as just an "untyped" token stream then.
